File: lib/userData.ts
import fs from 'fs';
import path from 'path';

const DATA_FILE_PATH = path.join(process.cwd(), 'data', 'user_data.json');

export interface Activity {
  id: string;
  action: string;
  target: string;
  timestamp: string;
}

export interface UserData {
  username?: string;
  email?: string;
  joinDate: string | null;
  totalTime: number; // in seconds
  avatar: string | null; // base64
  signature?: string;
  activities: Activity[];
}

export function getUserData(): UserData {
  try {
    if (!fs.existsSync(DATA_FILE_PATH)) {
       const initialData: UserData = {
         username: "Administrator",
         email: "admin@tideoa.com",
         joinDate: new Date().toISOString(), // First access sets join date
         totalTime: 0,
         avatar: null,
         signature: "Per aspera ad astra.",
         activities: []
       };
       saveUserData(initialData);
       return initialData;
    }
    const data = fs.readFileSync(DATA_FILE_PATH, 'utf-8');
    return JSON.parse(data);
  } catch (error) {
    console.error("Error reading user data:", error);
    return { username: "Administrator", email: "admin@tideoa.com", joinDate: null, totalTime: 0, avatar: null, signature: "System User", activities: [] };
  }
}

export function saveUserData(data: UserData) {
  try {
    fs.writeFileSync(DATA_FILE_PATH, JSON.stringify(data, null, 2), 'utf-8');
  } catch (error) {
    console.error("Error writing user data:", error);
  }
}

export function logUserActivity(action: string, target: string) {
  const data = getUserData();
  const newActivity: Activity = {
    id: Date.now().toString(),
    action,
    target,
    timestamp: new Date().toISOString()
  };
  // Keep only last 50 activities
  data.activities = [newActivity, ...data.activities].slice(0, 50);
  saveUserData(data);
}

File: backend/services/user_data.py
import json
import os
from datetime import datetime
from backend.models import UserData, Activity

# Assuming running from root or backend, adjust path accordingly
# In Next.js it was path.join(process.cwd(), 'data', 'user_data.json')
# We'll assume the backend is run from project root or handle relative path.
# Let's fix path to be relative to the backend execution or absolute.

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_FILE_PATH = os.path.join(BASE_DIR, 'data', 'user_data.json')

def get_user_data() -> UserData:
    try:
        if not os.path.exists(DATA_FILE_PATH):
            initial_data = UserData(
                joinDate=datetime.now().isoformat()
            )
            save_user_data(initial_data)
            return initial_data
        
        with open(DATA_FILE_PATH, 'r', encoding='utf-8') as f:
            data = json.load(f)
            return UserData(**data)
    except Exception as e:
        print(f"Error reading user data: {e}")
        return UserData(joinDate=None)

def save_user_data(data: UserData):
    try:
        os.makedirs(os.path.dirname(DATA_FILE_PATH), exist_ok=True)
        with open(DATA_FILE_PATH, 'w', encoding='utf-8') as f:
            f.write(data.model_dump_json(indent=2))
    except Exception as e:
        print(f"Error writing user data: {e}")

def log_user_activity(action: str, target: str):
    data = get_user_data()
    new_activity = Activity(
        id=str(int(datetime.now().timestamp() * 1000)),
        action=action,
        target=target,
        timestamp=datetime.now().isoformat()
    )
    # Keep only last 50 activities
    data.activities = [new_activity] + data.activities
    data.activities = data.activities[:50]
    save_user_data(data)
